crates/scroll/node/src/node.rs

```diff
@@ -0,0 +1,333 @@
+
+use alloy_rpc_types_engine::{ExecutionPayload, ExecutionPayloadSidecar, PayloadError};
+use reth_consensus::noop::NoopConsensus;
+use reth_db::transaction::{DbTx, DbTxMut};
+use reth_ethereum_engine_primitives::{EthEngineTypes, EthPayloadAttributes};
+use reth_ethereum_forks::EthereumHardforks;
+use reth_evm::execute::BasicBlockExecutorProvider;
+use reth_network::{NetworkHandle, PeersInfo};
+use reth_node_builder::{
+    components::{
+        ComponentsBuilder, ConsensusBuilder, ExecutorBuilder, NetworkBuilder,
+        PayloadServiceBuilder, PoolBuilder,
+    },
+    rpc::{EngineValidatorBuilder, RpcAddOns},
+    AddOnsContext, BuilderContext, EngineApiMessageVersion, EngineObjectValidationError,
+    EngineTypes, EngineValidator, FullNodeComponents, FullNodeTypes, Node, NodeAdapter,
+    NodeComponentsBuilder, PayloadOrAttributes,
+};
+use reth_node_types::{NodeTypes, NodeTypesWithDB, NodeTypesWithEngine};
+use reth_payload_builder::PayloadBuilderHandle;
+use reth_primitives::{Block, BlockBody, EthPrimitives, SealedBlock};
+use reth_provider::{
+    providers::ChainStorage, BlockBodyReader, BlockBodyWriter, ChainSpecProvider, DBProvider,
+    EthStorage, ProviderResult, ReadBodyInput,
+};
+use reth_rpc::EthApi;
+use reth_scroll_chainspec::ScrollChainSpec;
+use reth_scroll_evm::{ScrollEvmConfig, ScrollExecutionStrategyFactory};
+use reth_scroll_state_commitment::BinaryMerklePatriciaTrie;
+use reth_tracing::tracing::info;
+use reth_transaction_pool::{noop::NoopTransactionPool, TransactionPool};
+
+// #################### NODE ####################
+
+/// The Scroll node implementation.
+#[derive(Clone, Debug)]
+pub struct ScrollNode;
+
+impl<N> Node<N> for ScrollNode
+where
+    N: FullNodeTypes,
+    N::Types: NodeTypesWithDB
+        + NodeTypesWithEngine<
+            ChainSpec = ScrollChainSpec,
+            Primitives = EthPrimitives,
+            Engine = EthEngineTypes,
+            Storage = ScrollStorage,
+        >,
+{
+    type ComponentsBuilder = ComponentsBuilder<
+        N,
+        ScrollPoolBuilder,
+        ScrollPayloadBuilder,
+        ScrollNetworkBuilder,
+        ScrollExecutorBuilder,
+        ScrollConsensusBuilder,
+    >;
+
+    type AddOns = ScrollAddOns<
+        NodeAdapter<N, <Self::ComponentsBuilder as NodeComponentsBuilder<N>>::Components>,
+    >;
+
+    fn components_builder(&self) -> Self::ComponentsBuilder {
+        ComponentsBuilder::default()
+            .node_types::<N>()
+            .pool(ScrollPoolBuilder)
+            .payload(ScrollPayloadBuilder)
+            .network(ScrollNetworkBuilder)
+            .executor(ScrollExecutorBuilder)
+            .consensus(ScrollConsensusBuilder)
+    }
+
+    fn add_ons(&self) -> Self::AddOns {
+        ScrollAddOns::default()
+    }
+}
+
+// #################### NODE TYPES ####################
+
+impl NodeTypesWithEngine for ScrollNode {
+    type Engine = EthEngineTypes;
+}
+
+impl NodeTypes for ScrollNode {
+    type Primitives = EthPrimitives;
```

Comment:
Needed to switch to the `EthPrimitives` structure because of the following [trait bound](https://github.com/paradigmxyz/reth/blob/735eb4b97cf3937886a2972647e0afb90c916a71/crates/cli/commands/src/prune.rs#L19), where [`CliNodeTypes`](https://github.com/paradigmxyz/reth/blob/26fc701814e4c35abe5a5c8dfd285ca66c5d9255/crates/cli/commands/src/common.rs#L199) imposes the `Primitives = EthPrimitives`. This could cause some issue in the future if we ever need to define our own types.